compiler/functions.go

```diff
@@ -0,0 +1,320 @@
+package compiler
+
+// functions.go contains logic responsible for translating top-level functions
+// and function literals.
+
+import (
+	"bytes"
+	"fmt"
+	"go/ast"
+	"go/types"
+	"sort"
+	"strings"
+
+	"github.com/gopherjs/gopherjs/compiler/analysis"
+	"github.com/gopherjs/gopherjs/compiler/astutil"
+	"github.com/gopherjs/gopherjs/compiler/internal/typeparams"
+	"github.com/gopherjs/gopherjs/compiler/typesutil"
+)
+
+// newFunctionContext creates a new nested context for a function corresponding
+// to the provided info and instance.
+func (fc *funcContext) nestedFunctionContext(info *analysis.FuncInfo, sig *types.Signature, inst typeparams.Instance) *funcContext {
+	if info == nil {
+		panic(fmt.Errorf("missing *analysis.FuncInfo"))
```

Comment:
optional

```suggestion
		panic(errors.New("missing *analysis.FuncInfo"))
```

Reply:
I tend to use `fmt.Errorf()` even by default, because in practice it's likely at the error message would be extended at some point anyway and I'd have to update the function. `errors.New()` is really only worth it when you don't want to depend on the `fmt` package for some reason, and our code already depends on it.